command/job_restart.go

```diff
@@ -0,0 +1,849 @@
+	var batchWaitStr string
+	var groups []string
+	var tasks []string
+
+	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
+	flags.Usage = func() { c.ui.Output(c.Help()) }
+	flags.BoolVar(&c.allTasks, "all-tasks", false, "")
+	flags.StringVar(&batchSizeStr, "batch-size", "1", "")
+	flags.StringVar(&batchWaitStr, "batch-wait", "0s", "")
+	flags.BoolVar(&c.failOnError, "fail-on-error", false, "")
+	flags.BoolVar(&c.noShutdownDelay, "no-shutdown-delay", false, "")
+	flags.BoolVar(&c.reschedule, "reschedule", false, "")
+	flags.BoolVar(&c.verbose, "verbose", false, "")
+	flags.Var((funcVar)(func(s string) error {
+		groups = append(groups, s)
+		return nil
+	}), "group", "")
+	flags.Var((funcVar)(func(s string) error {
+		tasks = append(tasks, s)
+		return nil
+	}), "task", "")
+
+	err := flags.Parse(args)
+	if err != nil {
+		// Let the flags library handle and print the error message.
+		return nil, 1
+	}
+
+	// Truncate IDs unless full length is requested.
+	c.length = shortId
+	if c.verbose {
+		c.length = fullId
+	}
+
+	// Check that we got exactly one job.
+	args = flags.Args()
+	if len(args) != 1 {
+		return fmt.Errorf("This command takes one argument: <job>"), 1
+	}
+	c.jobID = strings.TrimSpace(args[0])
+
+	// Parse and validate -batch-size.
+	matches := jobRestartBatchSizeValueRegex.FindStringSubmatch(batchSizeStr)
+	if len(matches) != 2 {
+		return fmt.Errorf(
+			"Invalid -batch-size value %q: batch size must be an integer or a percentage",
+			batchSizeStr,
+		), 1
+	}
+
+	c.batchSizePercent = strings.HasSuffix(batchSizeStr, "%")
+	c.batchSize, err = strconv.Atoi(matches[1])
+	if err != nil {
+		return fmt.Errorf("Invalid -batch-size value %s: %v", batchSizeStr, err), 1
+	}
+	if c.batchSize == 0 {
+		return fmt.Errorf(
+			"Invalid -batch-size value %q: number value must be greater than zero",
+			batchSizeStr,
+		), 1
+	}
+
+	// Parse and validate -batch-wait.
+	if strings.ToLower(batchWaitStr) == jobRestartWaitAsk {
+		if !isTty() {
+			return fmt.Errorf(
+				"Invalid -batch-wait value %[1]q: %[1]q cannot be used when terminal is not interactive",
+				jobRestartWaitAsk,
+			), 1
+		}
+		c.batchWaitAsk = true
+	} else {
+		c.batchWait, err = time.ParseDuration(batchWaitStr)
+		if err != nil {
+			return fmt.Errorf("Invalid -batch-wait value %q: %v", batchWaitStr, err), 1
+		}
+	}
+
+	// -all-tasks conflicts with -task and <task>.
+	if c.allTasks && len(tasks) != 0 {
+		return fmt.Errorf("The -all-tasks option cannot be used with -task"), 1
+	}
+
+	// -reschedule conflicts with -task and <task>.
+	if c.reschedule && len(tasks) != 0 {
+		return fmt.Errorf("The -reschedule option cannot be used with -task"), 1
+	}
+
+	// Dedup tasks and groups.
+	c.groups = set.From(groups)
+	c.tasks = set.From(tasks)
+
+	return nil, 0
+}
+
+// validateGroupsAndTasks validates that the combination of groups and tasks
+// defined in the command are valid for the target job.
+func (c *JobRestartCommand) validateGroupsAndTasks(job *api.Job) error {
```

Comment:
Hum....good point. Your [previous comment](https://github.com/hashicorp/nomad/pull/16278#discussion_r1122364290) made me think about job versions and then I arrived at the same problem you described here.

I think it makes sense to not have this validation, but I think I need to fix the allocation filter logic to take the allocation's job version into consideration.

The only thing this would catch is if you pass a task or group that doesn't exist in any job version, but that's probably OK? If none of your groups or tasks matches then we will just end-up with a list of empty allocations to restart, which is fine as well.

I will refactor this validation and alloc filtering logic to be version aware, thanks!

Reply:
Refactored in https://github.com/hashicorp/nomad/pull/16278/commits/7c7e7483c237e8e01e3e09b3b9f736857bb5649d.